Fix issue where you can't buy a new Creator plan because the previously selected site already has one

mpkelly opened the pull request:
<img width="844" alt="Screenshot 2567-06-18 at 16 18 32" src="https://github.com/Automattic/wp-calypso/assets/6851384/d048c291-0369-4f07-b3db-2453d7ec117c">

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

I found this bug when testing a diff. To reproduce it you can do this:

1. Go to wordpress.com/sites and select a site with Creator
2. Click "My home"
3. Next, go to wordpress.com/hositng and click "Create a site"
4. Confirm you see disabled CTAs like above

The issue I found is that the previously selected site's plan is considered when deciding which buttons to show. This is not appropriate for `/setup/new-hosted-site` flow, which is creating a new site and should always be able to buy a plan. You can check my comments for more context. 

## Proposed Changes

* if we're on the `/setup/new-hosted-site` flow, don't load the current plan. If the current plan is empty, always allow new plan purchases in the grid. 


## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Reproduce the bug as above
* Apply patch and try and reproduce in your local Calypso environment
* I can't say I like this fix. Have a better idea?

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

mpkelly committed:
Fix issue where you can't buy a new Creator plan because the previously selected site already has one

zaguiini commented:
I think we can go with a simpler solution and `dispatch( setSelectedSite( null ) )` whenever someone enters the new/import hosted site flows.

chriskmnds commented on packages/plans-grid-next/src/components/actions.tsx:
```diff
@@ -97,6 +97,11 @@ const PlanFeatures2023GridActions = ( {
+	// availableForPurchase ultimately comes from use-check-plan-availability-for-purchase.ts
+	// which considers the plan of the current site. This is not appropriate for flows
+	// that are creating a new site, like /setup/new-hosted-site, so we override this value here.
+	availableForPurchase = availableForPurchase || ! currentSitePlanSlug;
```
Is it possible to incorporate this logic in `use-check-plan-availability-for-purchase.ts`? If we can extend the conditions there to account for this, then it sounds better than overriding. The same hook is used for pricing.

Let me know if I should help more / if it proves unwieldy.

p.s. shouldn't `|| ! currentSitePlanSlug` be already part of the first condition (below)? 🤔 

```
// 1. No selected site or current plan: plan is in a context-free state and is available for purchase
if ( ! selectedSiteId || ! currentPlan ) {
    return [ planSlug, true ];
}

// 2. Otherwise, check if the plan is available for purchase on the selected site
return [ planSlug, isPlanAvailableForPurchase( state, selectedSiteId, planSlug ) ];
```

mpkelly replied:
I changed the approach, which fixes it for the hosting flows. I don't think it affects any others, so this change should be enough, but it would be good to double-check.

mpkelly committed:
Revert previous change

mpkelly committed:
Reset the current site when starting hosting flows

mpkelly commented:
> dispatch( setSelectedSite( null ) )

Thanks, @zaguiini. That's a better solution. I've updated. I need to find out why we have a `NEW_HOSTED_SITE_FLOW_USER_INCLUDED` but as it's mostly the same code I updated it.

vykes-mac reviewed (approved):
I say we merge as is while looking for a more elegant solution
